src/main/scala/loamstream/model/execute/RxExecuter.scala

```diff
@@ -237,6 +234,8 @@ object RxExecuter extends Loggable {
     addExecutionRecorder(addJobFilter(default))
   }
   
+  private[execute] def deDupe(jobRuns: Observable[JobRun]): Observable[JobRun] = jobRuns.distinct(_.key)
```

Comment:
I don't know how much this matters, but my concern here would be job states not being taken into account. The `distinct` function (to my knowledge) takes the "first" value in the collection and drops all the others that command equally to it. So, for example, if you had:

```
val jobs = Seq(
  JobRun(1, NotStarted, 2), 
  JobRun(1, Running, 2)
)
```

Then after `distinct` you'd end up with `Seq(JobRun(1, NotStarted, 2))`, which may not be the desired output. I'd think you'd want to keep the jobs around with the "furthest" state progress? So, something on the order of...

```
NotStarted < Skipped < Submitted < Running < Completed < Failed
```

Then you can change the code to `jobRuns.sortBy(_.key).distinct(_.key)`.

All of the above may be moot. Perhaps all that matters is the ID of the job, but I just want to make sure that the de-duped list isn't (currently or accidentally later) used for waiting, starting more jobs, etc. and then causing duplicate work to happen multiple times, waiting longer than necessary/infinitely, etc.

Reply:
These are good points.  I've been buried in this for 1.5 weeks, so I definitely left out some context.

> the "furthest" state 

I thought about this, and have tried to implement some notion of an ordering for states in the past.  (There sort of is one now, in the form of terminal and non-terminal states.) The tricky part is that now those states are used to surface state information from the underlying infrastructure (Google/Uger/LSF), a la `Queued`, `Submitted`, etc, and those things can move jobs between states such that a natural ordering probably isn't possible.  I tried representing state transitions as a state machine to validate state transitions, but never wound up using it, since for non-local jobs, the state is whatever the infrastructure says it is.

In my failed branch, I didn't surface the states from the underlying infrastructure at all.  They got logged, but `ChunkRunner`s no longer produced a stream of states for a job, but instead a future for each job that completes once with a terminal states, or fails.  This part was good, and worth salvaging.  It simplified job-running, and allows for a natural ordering of states.

(Note that you can't sort a potentially-infinite stream of values emitted by an `Observable` without buffering them all up, but I get your point.)

Back to "the latest JobRun wins", you make a fair point about the statuses.  There are several considerations:
* You can't just de-dupe based on job id, since the same job may become runnable many times, say if it fails and LS's config leads to it being re-run.  (This is the reason `JobRun` exists at all.)
* I wanted something "aggressive" that would throw out more duplicates rather than less, given the spookiness of the bug (I was getting the feeling we all dread: "how did this ever work"?), and because in practice, I was only ever seeing duplicates of the same state transition (`JobRun`).  (There is reason to believe that's guaranteed.)

I'll try a test run where de-duping includes statuses.  It's easy to trigger the bad behavior, so I'll know quickly if it works or not.  If the run still proceeds correctly, I'll push a commit that includes statuses in the de-duping process.  If nothing else, it will make more sense in the future.

Reply:
> I'll try a test run where de-duping includes statuses. It's easy to trigger the bad behavior, so I'll know quickly if it works or not. If the run still proceeds correctly, I'll push a commit that includes statuses in the de-duping process. If nothing else, it will make more sense in the future.

I tried considering job statuses when de-duping, and it worked just as well as before: the unit and integration tests passed, resource usage was the same, and a run with the BioMe config is showing now pathological behavior after a couple of hours.  I pushed the relevant commits.  What do you think, @massung ?

Reply:
Looks good.